templates/commands/render.go

```diff
@@ -367,76 +366,123 @@ func parseAndExecuteGoTmpl(m model.String, inputs map[string]string) (string, er
 }
 
 // "from" may be a file or directory. "pos" is only used for error messages.
-func copyRecursive(pos *model.ConfigPos, from, to string, rfs renderFS, overwrite bool) error {
-	return fs.WalkDir(rfs, from, func(path string, d fs.DirEntry, err error) error { //nolint:wrapcheck
+func copyRecursive(pos *model.ConfigPos, fromRoot, toRoot string, rfs renderFS, overwrite, dryRun bool) error {
+	return fs.WalkDir(rfs, fromRoot, func(path string, de fs.DirEntry, err error) error { //nolint:wrapcheck
 		if err != nil {
 			return err // There was some filesystem error. Give up.
 		}
 
 		// We don't have to worry about symlinks here because we passed
 		// DisableSymlinks=true to go-getter.
 
-		relToSrc, err := filepath.Rel(from, path)
-		if err != nil {
-			return model.ErrWithPos(pos, "filepath.Rel(%s,%s): %w", from, path, err) //nolint:wrapcheck
+		if de.IsDir() {
+			return copyRecursiveDir(path, de, pos, fromRoot, toRoot, rfs, dryRun)
 		}
 
-		dst := filepath.Join(to, relToSrc)
+		return copyRecursiveFile(path, de, pos, fromRoot, toRoot, rfs, overwrite, dryRun)
+	})
+}
 
-		// The spec file may specify a file to copy that's deep in a
-		// directory tree, without naming its parent directory. We can't
-		// rely on WalkDir having traversed the parent directory of $path,
-		// so we must create the target directory if it doesn't exist.
-		dirToCreate := dst
-		if !d.IsDir() {
-			dirToCreate = filepath.Dir(dst)
-		}
+// A helper function that's called by the WalkDirFunc inside copyRecursive in
+// the case that the source path is a directory.
+func copyRecursiveDir(path string, de fs.DirEntry, pos *model.ConfigPos, fromRoot, toRoot string, rfs renderFS, dryRun bool) error {
```

Comment:
Dumb question - since we already create parent dir when copying files, can we just skip copying dirs entirely? 
One obvious caveat is that we then can't template empty folders.

Reply:
But considering git also ignore empty folders, maybe that's actually good.

Reply:
> since we already create parent dir when copying files,

This is only sometimes true. In the general case, we can't assume that the parent has been traversed. The user may do something like this:

```
- action: include
  params:
    paths: ['a/deeply/nested/file.txt']
```

In which case the directories "a", "deeply", and "nested" will need to be created by mkdirall, and will NOT be seen by the WalkDirFunc. So I think the logic is corrrect as written.

 I tried to make this clear in the comment at line 428. LMK if you think it needs to be reworded.